### PR TITLE
Build the runtime with -ftls-model=initial-exec where available

### LINKED ISSUE #3168: tlsglobals crash with SMP and shared libraries
See #3166 for context.

`./build LIBS netlrts-linux-x86_64 smp --build-shared --with-production` crashes in the `cxx-tlsglobals` test:

- Removing `smp` or `--build-shared` works around this issue.
- buildold is also affected.

```
../../../bin/testrun  ./cxx-tlsglobals +p2 +vp4 +balancer RandCentLB ++local +setcpuaffinity
Charmrun> scalable start enabled. 
Charmrun> started all node programs in 0.022 seconds.
Charm++> Running in SMP mode: 2 processes, 1 worker threads (PEs) + 1 comm threads per process, 2 PEs total
Charm++> The comm. thread both sends and receives messages
Charm++> -tlsglobals enabled for privatization of thread-local variables.
Converse/Charm++ Commit ID: 408824f
Isomalloc> Synchronized global address space.
Charm++> scheduler running in netpoll mode.
CharmLB> Load balancer assumes all CPUs are same.
Charm++> cpu affinity enabled. 
Charm++> Running on 1 hosts (1 sockets x 2 cores x 1 PUs = 2-way SMP)
Charm++> cpu topology info is gathered in 0.100 seconds.
Charm++> Warning: Running with more SMP threads (4) than physical cores (2).
         Use +CmiSleepOnIdle (default behavior) or +CmiSpinOnIdle to silence this message.
[0] TreeLB in LEGACY MODE support
[0] TreeLB: Using PE_Root tree with: Random 
Beginning round of testing.
Test #01: extern global, in same object
#01 - [0](0) - 0x000400000490 - single write - passed
#01 - [1](0) - 0x155750000490 - single write - passed
#01 - [2](1) - 0x2aaaa0000490 - single write - passed
#01 - [3](1) - 0x3ffdf0000490 - single write - passed
#01 - [0](0) - 0x000400000490 - many write - passed
#01 - [1](0) - 0x155750000490 - many write - passed
#01 - [2](1) - 0x2aaaa0000490 - many write - passed
#01 - [3](1) - 0x3ffdf0000490 - many write - passed
Test #02: static global, in same object
#02 - [0](0) - 0x00040000048c - single write - passed
#02 - [1](0) - 0x15575000048c - single write - passed
#02 - [2](1) - 0x2aaaa000048c - single write - passed
#02 - [3](1) - 0x3ffdf000048c - single write - passed
#02 - [0](0) - 0x00040000048c - many write - passed
#02 - [1](0) - 0x15575000048c - many write - passed
#02 - [2](1) - 0x2aaaa000048c - many write - passed
#02 - [3](1) - 0x3ffdf000048c - many write - passed
Test #03: static local, in same object
#03 - [0](0) - 0x000400000488 - single write - passed
#03 - [1](0) - 0x155750000488 - single write - passed
#03 - [2](1) - 0x2aaaa0000488 - single write - passed
#03 - [3](1) - 0x3ffdf0000488 - single write - passed
#03 - [0](0) - 0x000400000488 - many write - passed
#03 - [1](0) - 0x155750000488 - many write - passed
#03 - [2](1) - 0x2aaaa0000488 - many write - passed
#03 - [3](1) - 0x3ffdf0000488 - many write - passed
Test #04: extern global, in other object
#04 - [0](0) - 0x00040000049c - single write - passed
#04 - [1](0) - 0x15575000049c - single write - passed
#04 - [2](1) - 0x2aaaa000049c - single write - passed
#04 - [3](1) - 0x3ffdf000049c - single write - passed
#04 - [0](0) - 0x00040000049c - many write - passed
#04 - [1](0) - 0x15575000049c - many write - passed
#04 - [2](1) - 0x2aaaa000049c - many write - passed
#04 - [3](1) - 0x3ffdf000049c - many write - passed
Test #05: static global, in other object
#05 - [0](0) - 0x000400000498 - single write - passed
#05 - [1](0) - 0x155750000498 - single write - passed
#05 - [2](1) - 0x2aaaa0000498 - single write - passed
#05 - [3](1) - 0x3ffdf0000498 - single write - passed
#05 - [0](0) - 0x000400000498 - many write - passed
#05 - [1](0) - 0x155750000498 - many write - passed
#05 - [2](1) - 0x2aaaa0000498 - many write - passed
#05 - [3](1) - 0x3ffdf0000498 - many write - passed
Test #06: static local, in other object
#06 - [0](0) - 0x000400000494 - single write - passed
#06 - [1](0) - 0x155750000494 - single write - passed
#06 - [2](1) - 0x2aaaa0000494 - single write - passed
#06 - [3](1) - 0x3ffdf0000494 - single write - passed
#06 - [0](0) - 0x000400000494 - many write - passed
#06 - [1](0) - 0x155750000494 - many write - passed
#06 - [2](1) - 0x2aaaa0000494 - many write - passed
#06 - [3](1) - 0x3ffdf0000494 - many write - passed
Test #07: extern global, in static library
#07 - [0](0) - 0x0004000004cc - single write - passed
#07 - [1](0) - 0x1557500004cc - single write - passed
#07 - [2](1) - 0x2aaaa00004cc - single write - passed
#07 - [3](1) - 0x3ffdf00004cc - single write - passed
#07 - [0](0) - 0x0004000004cc - many write - passed
#07 - [1](0) - 0x1557500004cc - many write - passed
#07 - [2](1) - 0x2aaaa00004cc - many write - passed
#07 - [3](1) - 0x3ffdf00004cc - many write - passed
Test #08: static global, in static library
#08 - [0](0) - 0x0004000004c8 - single write - passed
#08 - [1](0) - 0x1557500004c8 - single write - passed
#08 - [2](1) - 0x2aaaa00004c8 - single write - passed
#08 - [3](1) - 0x3ffdf00004c8 - single write - passed
#08 - [0](0) - 0x0004000004c8 - many write - passed
#08 - [1](0) - 0x1557500004c8 - many write - passed
#08 - [2](1) - 0x2aaaa00004c8 - many write - passed
#08 - [3](1) - 0x3ffdf00004c8 - many write - passed
Test #09: static local, in static library
#09 - [0](0) - 0x0004000004c4 - single write - passed
#09 - [1](0) - 0x1557500004c4 - single write - passed
#09 - [2](1) - 0x2aaaa00004c4 - single write - passed
#09 - [3](1) - 0x3ffdf00004c4 - single write - passed
#09 - [0](0) - 0x0004000004c4 - many write - passed
#09 - [1](0) - 0x1557500004c4 - many write - passed
#09 - [2](1) - 0x2aaaa00004c4 - many write - passed
#09 - [3](1) - 0x3ffdf00004c4 - many write - passed
Test #10: extern global, in shared library
#10 - [0](0) - 0x000400000478 - single write - passed
#10 - [1](0) - 0x155750000478 - single write - passed
#10 - [2](1) - 0x2aaaa0000478 - single write - passed
#10 - [3](1) - 0x3ffdf0000478 - single write - passed
#10 - [0](0) - 0x000400000478 - many write - passed
#10 - [1](0) - 0x155750000478 - many write - passed
#10 - [2](1) - 0x2aaaa0000478 - many write - passed
#10 - [3](1) - 0x3ffdf0000478 - many write - passed
Test #11: static global, in shared library
#11 - [0](0) - 0x000400000474 - single write - passed
#11 - [1](0) - 0x155750000474 - single write - passed
#11 - [2](1) - 0x2aaaa0000474 - single write - passed
#11 - [3](1) - 0x3ffdf0000474 - single write - passed
#11 - [0](0) - 0x000400000474 - many write - passed
#11 - [1](0) - 0x155750000474 - many write - passed
#11 - [2](1) - 0x2aaaa0000474 - many write - passed
#11 - [3](1) - 0x3ffdf0000474 - many write - passed
Test #12: static local, in shared library
#12 - [0](0) - 0x000400000470 - single write - passed
#12 - [1](0) - 0x155750000470 - single write - passed
#12 - [2](1) - 0x2aaaa0000470 - single write - passed
#12 - [3](1) - 0x3ffdf0000470 - single write - passed
#12 - [0](0) - 0x000400000470 - many write - passed
#12 - [1](0) - 0x155750000470 - many write - passed
#12 - [2](1) - 0x2aaaa0000470 - many write - passed
#12 - [3](1) - 0x3ffdf0000470 - many write - passed
Round of testing complete.
Requesting migration.
[1] About to migrate.
[3] About to migrate.
Beginning round of testing.
Test #13: extern global, in same object
------------- Processor 0 Exiting: Caught Signal ------------
Reason: segmentation violation
Suggestion: Try running with '++debug', or linking with '-memory paranoid' (memory paranoid requires '+netpoll' at runtime).
[0] Stack Traceback:
  [0:0] cxx-tlsglobals 0x55555584b1df 
  [0:1] libpthread.so.0 0x7ffff79c4980 
  [0:2] cxx-tlsglobals 0x555555678bfa MPI_Comm_rank
  [0:3] cxx-tlsglobals 0x55555565f143 
  [0:4] cxx-tlsglobals 0x555555675975 ampiParent::resumeAfterMigration()
  [0:5] cxx-tlsglobals 0x5555556759dc CkIndex_ampiParent::_call_resumeAfterMigration_void(void*, void*)
  [0:6] cxx-tlsglobals 0x555555753bb4 CkDeliverMessageFree
  [0:7] cxx-tlsglobals 0x55555579c912 CkLocRec::invokeEntry(CkMigratable*, void*, int, bool)
  [0:8] cxx-tlsglobals 0x55555575bc49 _processHandler(void*, CkCoreState*)
  [0:9] cxx-tlsglobals 0x55555580462a CsdScheduleForever
  [0:10] cxx-tlsglobals 0x5555558048a5 CsdScheduler
  [0:11] cxx-tlsglobals 0x55555584d7ba 
  [0:12] cxx-tlsglobals 0x55555584eb84 ConverseInit
  [0:13] cxx-tlsglobals 0x5555557f058c charm_main
  [0:14] libc.so.6 0x7ffff6837bf7 __libc_start_main
  [0:15] cxx-tlsglobals 0x55555565eb6a _start
Fatal error on PE 0> segmentation violation
```

### PR DESCRIPTION
**Pull request by evan-charmworks**
Fixes #3168

### CODE REVIEW

**Comment by evan-charmworks** [comment]
Not sure why this is happening:

```
CMake Error at cmake/detect-features-c.cmake:148 (check_c_compiler_flag):
  Unknown CMake command "check_c_compiler_flag".
Call Stack (most recent call first):
  cmake/detect-features.cmake:1 (include)
  CMakeLists.txt:434 (include)

-- Configuring incomplete, errors occurred!
```

File: cmake/detect-features-c.cmake [hunk 1]
```diff
@@ -144,6 +144,13 @@ check_function_exists(sysctlbyname HAVE_SYSCTLBYNAME)
 check_function_exists(uname HAVE_UNAME)
 check_function_exists(usleep CMK_HAS_USLEEP)
 
+# C compiler flags
```
Comment (matthiasdiener) [inline cmake/detect-features-c.cmake:147]:
```suggestion
# C compiler flags
include(CheckCCompilerFlag)
```

**Review by matthiasdiener — APPROVED** [review]

**Review by matthiasdiener — DISMISSED** [review]

**Review by matthiasdiener — APPROVED** [review]
Passes the spack test without workaround now.

**Comment by evan-charmworks** [comment]
Needs one more review.

**Review by stwhite91 — APPROVED** [review]

**Pull request merged**